GPflow/param.py

```diff
@@ -519,54 +526,118 @@ class AutoFlow:
     >>> with m.tf_mode():
     >>>     graph = tf.foo(m.baz, x_tf, y_tf)
-    >>> result = m._session.run(graph,
-                                feed_dict={x_tf:x,
-                                           y_tf:y,
-                                           m._free_vars:m.get_free_state()})
-
+    >>> result = m.session.run(graph, feed_dict={
+                     x_tf:x, y_tf:y, m._free_vars:m.get_free_state()})
     Not only is the syntax cleaner, but multiple calls to the method will
     result in the graph being constructed only once.
 
     """
 
     def __init__(self, *tf_arg_tuples):
-        # NB. TF arg_tuples is a list of tuples, each of which can be used to
-        # construct a tf placeholder.
+        """
+        :param tf_arg_tuples: list of tuples, each of which can be
+                              used to construct a tensorflow placeholder.
+        """
         self.tf_arg_tuples = tf_arg_tuples
 
     def __call__(self, tf_method):
-        @wraps(tf_method)
-        def runnable(instance, *np_args):
-            storage_name = '_' + tf_method.__name__ + '_AF_storage'
+        @functools.wraps(tf_method)
+        def runnable(instance, *np_args, **kwargs):
+            """
+            AutoFlow function wrapper adds extra parameters:
+            session and graph. It allows you specify which session and graph
+            will be used during compilation. Session always prevails graph
+            presence. It means that whenever session parameter is not None,
+            the session's graph will be used for compilation.
+            Let's say we have `method(x)` wrapped in AutoFlow:
+
+            > m.method(x)
+            > m.method(x, session=tf.Session())
+            > m.method(x, graph=tf.Graph())
+
+            In first case we call method without extra parameters -
+            the default graph is used and new session is created for compiling
+            model, both are stored in model cache.
+            Second example has only session parameter which is constructed
+            with default graph, thereafter model will not be re-built, but it
+            will be initialized again and new session will replace old one.
+            In third example we passed a new graph and the model will be
+            recompiled from scratch.
+            The graph and session being passed together make little sense as
+            tensorflow session can be tied to only graph.
+
+            :raises ValueError: when `**kwargs` contains unknown key-valued
+            parameters or user passed both session and graph parameters and
+            session references to different graph.
+            """
+
+            if not set(kwargs.keys()).issubset(['session', 'graph']):
+                raise ValueError('Unknown arguments passed.')
+
+            session = kwargs.get('session')
+            graph = kwargs.get('graph')
+
+            if session and graph and session.graph != graph:
```

Comment:
perhaps: `if (session and graph) and (graph is not session.graph)`  -- same logic but clearer?